### ISSUE TITLE
Misleading error message when keyword is used as a cte name

### ISSUE THREAD

**Issue by olexanderos** [sqlfluff/sqlfluff#7121]
### Search before asking

- [x] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

I have a dbt model that looks similar to this:
```sql
with src as (
    select ...
    from {{ source("my_source", "my_table") }}
),
extended as (
    select ...
    from src
)
select ...
from extended
```
The error message I get when I run `sqlfluff lint my_model.sql` is:
```
L:   1 | P:   1 |  PRS | Line 1, Position 1: Found unparsable section: 'with\n   
                       | source as (\n        select\n    ...'
WARNING: Parsing errors found and dialect is set to 'athena'. Have you configured your dialect correctly?
```

This is a misleading message and it is quite hard to find a rootcause.

### Expected Behaviour

Expected error message:
```
Keywords should not be used as identifiers. Used keyword: 'extended'
```

### Observed Behaviour

The error message:
```
L:   1 | P:   1 |  PRS | Line 1, Position 1: Found unparsable section: 'with\n   
                       | source as (\n        select\n    ...'
WARNING: Parsing errors found and dialect is set to 'athena'. Have you configured your dialect correctly?
```

### How to reproduce

1. Configure `athena` as sql dialect
2. Create a sql with cte name `extended`
3. run `sqlfluff lint your_sql.sql`

### Dialect

athena

### Version

3.3.1

### Configuration

```toml
[tool.sqlfluff.core]
exclude_rules = [
    "layout.indent",
    "layout.cte_bracket",
    "layout.cte_newline",
    "layout.select_targets",
    "layout.keyword_newline",
    "layout.spacing",
    "structure.using",
    "structure.column_order",
]
templater = "dbt"
dialect = "athena"

[tool.sqlfluff.rules]
capitalisation_policy = "lower"
extended_capitalisation_policy = "lower"

[tool.sqlfluff.rules.convention.terminator]
multiline_newline = true

[tool.sqlfluff.templater.jinja]
apply_dbt_builtins = true
```

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [x] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by Jyriu** [comment]
Hi! I'd like to work on this issue as my first contribution to sqlfluff. Is it still available to pick up?
